src/librustc_mir/borrow_check/move_errors.rs

```diff
@@ -257,81 +252,95 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
+                            let closure_kind = closure_kind_ty.to_opt_closure_kind();
+                            let place_description = match closure_kind {
+                                Some(ty::ClosureKind::Fn) => {
+                                    "captured variable in an `Fn` closure"
+                                }
+                                Some(ty::ClosureKind::FnMut) => {
+                                    "captured variable in an `FnMut` closure"
+                                }
+                                Some(ty::ClosureKind::FnOnce) => {
+                                    bug!("closure kind does not match first argument type")
+                                }
+                                None => bug!("closure kind not inferred by borrowck"),
+                            };
+                            debug!("report: closure_kind_ty={:?} closure_kind={:?} \
+                                    place_description={:?}", closure_kind_ty, closure_kind,
+                                    place_description);
+
+                            let mut diag = self.infcx.tcx.cannot_move_out_of(
+                                span, place_description, origin);
+
+                            for prefix in self.prefixes(&original_path, PrefixSet::All) {
+                                if let Some(field) = self.is_upvar_field_projection(prefix) {
+                                    let upvar_hir_id = self.upvars[field.index()].var_hir_id;
+                                    let upvar_span = self.infcx.tcx.hir().span_by_hir_id(
+                                        upvar_hir_id);
+                                    diag.span_label(upvar_span, "captured outer variable");
+                                    break;
                                 }
-
-                                diag
                             }
-                            _ => {
-                                let source = self.borrowed_content_source(place);
-                                self.infcx.tcx.cannot_move_out_of(
-                                    span, &source.to_string(), origin
-                                )
-                            },
+
+                            diag
                         }
+                        _ => {
+                            let source = self.borrowed_content_source(place);
+                            self.infcx.tcx.cannot_move_out_of(
+                                span, &source.to_string(), origin
+                            )
+                        },
+                    };
+                    let orig_path_ty = format!(
+                        "{:?}",
+                        original_path.ty(self.mir, self.infcx.tcx).ty,
+                    );
+                    let snippet = self.infcx.tcx.sess.source_map().span_to_snippet(span).unwrap();
+                    let is_option = orig_path_ty.starts_with("std::option::Option");
```

Comment:
How do you feel about waiting a bit with merging this PR (a bit being until https://github.com/rust-lang/rust/pull/60966 is merged)? Then you can annotate `Option` and `Result` with `#[rustc_diagnostic_item = "option"]` and `"result"` respectively and then check for these items via

```rust
if let Some(adt) = original_path.ty(self.mir, self.infcx.tcx).adt_def() {
	if self.infcx.tcx.is_diagnostic_item(adt.did, sym::option) {
        ...
    }
}
```

Reply:
There are a couple of places where we're already doing string comparisons like this, so I love the fact that we'll be able to get rid of them soon!

Reply:
Hmm... or we can just merge this and fix it up later, doesn't seem like a big deal in fact.

Reply:
I'm fine either way